activerecord/lib/active_record/relation/predicate_builder.rb

```diff
@@ -112,6 +112,12 @@ def create_binds_for_hash(attributes)
               binds.concat(bvs)
               attrs
             end
+          when value.is_a?(Array) && !table.type(column_name).respond_to?(:subtype)
```

Comment:
I agree that there's a problem with the structure.
How about the alternative #29033?
#29033 is a simple one line fix and we already have a similar code (for keeping existing expectation: https://github.com/rails/rails/blob/a6d6e90fa90eedbdb6d3cda679aa4b47ddfa7315/activerecord/lib/active_record/relation/predicate_builder/polymorphic_array_value.rb#L13).
I'd like to add the expectation that typical generated queries will be preparable even if association queries.